common/cache/src/disk_cache.rs

```diff
@@ -0,0 +1,361 @@
+// Copyright 2021 Datafuse Labs.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::boxed::Box;
+use std::ffi::OsStr;
+use std::ffi::OsString;
+use std::fs::File;
+use std::fs::{self};
+use std::hash::BuildHasher;
+use std::io;
+use std::io::prelude::*;
+use std::path::Path;
+use std::path::PathBuf;
+
+use filetime::set_file_times;
+use filetime::FileTime;
+use ritelinked::DefaultHashBuilder;
+use walkdir::WalkDir;
+
+use crate::Cache;
+use crate::FileSize;
+use crate::LruCache;
+
+/// Return an iterator of `(path, size)` of files under `path` sorted by ascending last-modified
+/// time, such that the oldest modified file is returned first.
+fn get_all_files<P: AsRef<Path>>(path: P) -> Box<dyn Iterator<Item = (PathBuf, u64)>> {
+    let mut files: Vec<_> = WalkDir::new(path.as_ref())
+        .into_iter()
+        .filter_map(|e| {
+            e.ok().and_then(|f| {
+                // Only look at files
+                if f.file_type().is_file() {
+                    // Get the last-modified time, size, and the full path.
+                    f.metadata().ok().and_then(|m| {
+                        m.modified()
+                            .ok()
+                            .map(|mtime| (mtime, f.path().to_owned(), m.len()))
+                    })
+                } else {
+                    None
+                }
+            })
+        })
+        .collect();
+    // Sort by last-modified-time, so oldest file first.
+    files.sort_by_key(|k| k.0);
+    Box::new(files.into_iter().map(|(_mtime, path, size)| (path, size)))
+}
+
+/// An LRU cache of files on disk.
+pub type LruDiskCache = DiskCache<LruCache<OsString, u64, DefaultHashBuilder, FileSize>>;
```

Comment:
We can easily define another disk cache based on a different policy by simply implementing the `common_cache::cache::Cache` trait.